### PR TITLE
[core] Sampling controller interface

### LINKED ISSUE #5423: [RFC]: Improve guided decoding (logit_processor) APIs and performance.
### Motivation.

Currently, guided decoding & logit processor API is incomplete has has several issues. The RFC is intended to bring up problems and solutions. Some of issues may have been already addressed and there are PRs out already.

There are 3 major issues.
- It is not supported from SamplingParamters
- It is not possible to support batch/async logit processing.
- Upon failures, engine will die.

### Proposed Change.

API
----
guided decoding parameters are not supported with SamplingParams. It is addressed from https://github.com/vllm-project/vllm/pull/4130

Performance
-------------
Currently, logit processors APIs are applied row by row blocking (https://github.com/vllm-project/vllm/blob/246598a6b1e22616630b7f1bf11bd9bcb31dc860/vllm/model_executor/layers/logits_processor.py#L112). Instead, we can use parallel processing (e.g., ray or thread pool) to improve the logit processing performance. We are using this mechanism internally at Anyscale. We'd like to support this feature in OSS, and would like to improve logit processor API to support 1. async. 2. batching. 

This requires logit processor to be
- stateful (to use a tool like Ray or thread pool). I think this PR https://github.com/vllm-project/vllm/pull/5329 is likely sufficient. 
- async. We'd like to propose "prepare" API which can separate out compute_logits from preparing logits.

```
class LogitPostProcessor:
   def initialize(self, logit_processor_config: LogitProcessorConfig):
       """Initialize the post processor. Post processor may have states
           such as thread pool or Ray actors. It should be initialized
           here.
       """
       ...

   def prepare(
           self,
           seq_gruop_metadata_list: List[SequenceGroupMetadata]):
       """Asynchronously prepare logit masks."""
       ...

   def apply(self, logits: torch.Tensor) -> torch.Tensor:
       """Apply the prepared masks to a given logits."""
       ...

# For each model, we will have

def compute_logits(...):
    ....

def prepare_logits(seq_group_metadata_list):
    ....

```

`prepare` and `apply` assume 1:1 calls. E.g., once prepare is called, apply has to be called before another prepare is called. I think it is the safe assumption. Alternatively, we can make prepare return a class, but that will make interface surface larger, so I don't prefer that solution (but I am open to hear feedback!)

This is the example usage of the API

```
        # each model will have prepare_logits API
        self.model.prepare_logits(seq_group_metadata_list)
        hidden_states = model_executable(
            input_ids=input_tokens,
            positions=input_positions,
            kv_caches=kv_caches,
            attn_metadata=attn_metadata,
            **multi_modal_kwargs,
        )
        # Compute the logits. logit processors are applied here.
        logits = self.model.compute_logits(hidden_states, sampling_metadata)
```

We are also considering to upstream Ray based batch processing implementation with lmformatenforcer.

Failure Handling
----------------
When using a stateful logit processor, it is possible requests are failed. For example, if we use Ray, Ray actors can die. Or there could be user's schema issue that cannot be caught ahead of time.

When it happens, we should fail the seq_group immediately. We will introduce a new status "FINISHED_INTERNAL_ERROR = enum.auto()" to https://github.com/vllm-project/vllm/blob/246598a6b1e22616630b7f1bf11bd9bcb31dc860/vllm/sequence.py#L42. If any logit processor is failed, we will mark the relevant seq_group as failed, and the request will be aborted. 

### Feedback Period.

_No response_

### CC List.

cc @simon-mo @Yard1 

### Any Other Things.

_No response_

### PR DESCRIPTION
**Pull request by mmoskal**
This patch adds `SamplingController` object on `LLMEngine`. It subsumes `LogitsProcessor` functions in `SamplingParams` as suggested in #5423. Instead of calling the class `LogitPostProcessor` I used `SamplingController` since it's a bit broader than just dealing with logits (in particular it allows for fast forward tokens, see below). I'm happy to rename if needed.

The basic idea is that the engine holds an instance of `SamplingController` and calls methods on it to influence the sampling process. In every step the following methods are called:
* `prepare(sampling_metadata: SamplingMetadata)`: this is meant to start computation of logit biases for sequences in the batch; the controller will likely need to store mapping from sequences to indices in logit tensor
* model forward pass is started
* `transform_logits(logits: torch.Tensor) -> torch.Tensor` is called on the entire logit tensor (for all sequences in the batch)
* sampling is performed as usual
* `transform_sampler_output(output: SamplerOutput) -> SamplerOutput` is called on the output of the sampler

In case of an empty step (where no sequences are scheduled to run), the `empty_step()` method is called, instead of the three methods mentioned above. This is to allow the controller to perform cleanup.

To be clear, the only way to use this right now, is to derive from `SamplingController` and use `vllm` as a library.

The transformation of sampler output is primary useful in conjunction with the newly added `SequenceOutput.fast_forward_tokens` field. If this is set, these tokens are to be added to the sequence instead of the sampled token.

An example, where fast forward tokens are useful is generating data adhering to a certain JSON schema. The controller first forces `{"name":"` to be generated, then the model generates `John"`, the controller forces `,\n"age":`, model generates `42`, and so on. Another example is chain-of-thought reasoning, where after the model generated a sentence, the controller forces more instructions for the model, the model generates more text, and so on. If used, these greatly speed up generation process.

The `SamplingController` is passed from `LLMEngine` to worker/executor via `ExecuteModelRequest` and `ModelRunnerInputBase`. Only driver worker receives the controller (as it's the only one that does sampling).

I also added `request_id` to `SequenceGroupToSample` (which is referenced from `SamplingMetadata`) to allow the controller to identify the sequences in the batch.

There are two places where I had to make changes to allow empty fast forward tokens:
- `SequenceData.update_num_computed_tokens`
- `BlockTable.append_token_ids`
The main reason to support empty tokens is for the cases when the computation of the next token mask did not complete in time. Instead of waiting for it to finish, or aborting the sequence, the controller can instead do an "empty pass" on that particular sequence and try it again in the next step, without holding up the rest of the batch. Of course, if this happens several times, the controller is free to terminate the sequence.

Status: I have this working with AICI. I think this is good to go, though it may need some tests.

CC @rkooo567 @simon-mo @GindaChen @cadedaniel @njhill 

FIX #5423

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by mmoskal** [comment]
@rkooo567 @simon-mo any comments?

**Comment by Peng-YM** [comment]
Looks great, thanks for your contribution!

**Comment by rkooo567** [comment]
I will take a look at it today!

**Comment by mmoskal** [comment]
@rkooo567 any updates?

File: vllm/core/block/block_table.py [hunk 1]
```diff
@@ -147,10 +147,13 @@ def append_token_ids(self,
 
         # Update the blocks with the new tokens
         first_block_idx = self._num_full_slots // self._block_size
-        token_blocks = self._chunk_token_blocks_for_append(token_ids)
 
-        for i, token_block in enumerate(token_blocks):
-            self._blocks.append_token_ids(first_block_idx + i, token_block)
+        # don't bother appending anything, if no new token_ids were generated
```
Comment (rkooo567) [inline vllm/core/block/block_table.py:151]: QQ: is this change related to this PR?

File: vllm/sequence.py [hunk 2]
```diff
@@ -912,6 +941,26 @@ def prune(self,
             self.seq_ids = seq_ids
 
 
+class SamplingController:
```
Comment (rkooo567) [inline vllm/sequence.py:944]: Make it an abstract class?

Comment (rkooo567) [reply vllm/sequence.py:944]: Also in the docstring, we should probably mention this class is a singleton and stateful. And prepare & transform logits need to be called 1 tot 1

File: vllm/sequence.py [hunk 3]
```diff
@@ -701,6 +707,29 @@ def __init__(
         self.parent_seq_id = parent_seq_id
         self.output_token = output_token
         self.logprobs = logprobs
+        # If present, these tokens should appended to the output
+        # instead of output_token.
+        self.fast_forward_tokens: Optional[List[int]] = None
+
+    def append_to(self, seq: Sequence) -> None:
```
Comment (rkooo567) [inline vllm/sequence.py:714]: can you add a docstring?

File: vllm/sequence.py [hunk 4]
```diff
@@ -468,8 +474,8 @@ def lora_int_id(self) -> int:
 
     def get_last_latency(self, now: float) -> Optional[float]:
         """Sets the last token time for Request level timings."""
-        # If still in prefill phase, raise Error.
-        if self.is_prefill():
+        # If still in initial prefill phase, raise Error.
```
Comment (rkooo567) [inline vllm/sequence.py:477]: what's the case it is not "initial"?

**Review by rkooo567 — COMMENTED** [review]
I think we will need review from some of more stakeholders. Maybe @njhill and @simon-mo  But the part I need looks pretty good to me. I am going to try prototyping our internal implementation by early next week and get back to you asap. 

I have 2 questions.
- For SamplingController, currently, there's no way to pass this. What's the API you are thinking to set this? (is it supposed to be set in a fork level?)
- For fast forward token, are you planning to upstream any implementation?

File: vllm/worker/model_runner.py [hunk 5]
```diff
@@ -1240,6 +1240,11 @@ def execute_model(
             "finished_requests_ids": model_input.finished_requests_ids,
             "request_ids_to_seq_ids": model_input.request_ids_to_seq_ids,
         } if self.has_seqlen_agnostic else {}
+
+        if (ctrl := model_input.sampling_controller) is not None:
+            assert model_input.sampling_metadata is not None
+            ctrl.prepare(model_input.sampling_metadata)
```
Comment (rkooo567) [inline vllm/worker/model_runner.py:1246]: This has been changed from RFC (we pass seq group metadata, here, we are using sampling params). Is there any way to hook this with seq group metadata?

File: vllm/core/block/block_table.py (hunk 1 above)
Comment (mmoskal) [reply vllm/core/block/block_table.py:151]: Yes, this happens when the "fast forward" is 0-tokens long - see comment in PR description about `BlockTable.append_token_ids`

File: vllm/sequence.py (hunk 4 above)
Comment (mmoskal) [reply vllm/sequence.py:477]: This happens for fast-forward tokens - if there is more than one, we switch to prefill mode.

File: vllm/worker/model_runner.py (hunk 5 above)
Comment (mmoskal) [reply vllm/worker/model_runner.py:1246]: There are `SamplingParams` in sampling_metadata (specifically in `SequenceGroupToSample`); there is also `request_id` and `sequence_id` - I think this is the stuff you said you needed. The main reason to use sampling metadata is that it also has all the logit positions already computed.

Comment (rkooo567) [reply vllm/worker/model_runner.py:1246]: I see. yeah I guess what we have in sampling metdata is probably sufficient. let me get back to you soon about this.

File: vllm/sequence.py (hunk 2 above)
Comment (mmoskal) [reply vllm/sequence.py:951]: Added docstrings.

It seems python doesn't have abstract classes, only abstract methods. I can imagine use cases where you only override some of these methods, and none of them you have to override, so I don't think it makes sense to make any of them abstract.

One thing we could do is to make the engine.sampling_controller field non-optional and use this base class as a no-op implementation. Not sure if that would be cleaner though?

**Comment by mmoskal** [comment]
Regarding usage pattern - it currently requires you to construct LLMEngine yourself. I plan to do a PR on the openai api_server.py to make it a bit more modular, so you can have your own api_server that extends the "openai" one. See here for an usage example: https://github.com/microsoft/aici/blob/split_toktrie/py/pyaici/vllm_server.py

Regarding fast-forward tokens - they work with the current patch, provided you enable chunked prefill. AICI supports them and so does the Low-level Guidance parser I've been working on (right now https://github.com/mmoskal/llguidance but it will be moving gh orgs shortly)

I pushed the docstring changes, and now it seems some tests are failing, not sure what that is about...

**Comment by rkooo567** [comment]
okay, let me prototype our internal changes with this API and I will get back to you asap. In the meantime let's wait for additional community reviews. other than that, it looks okay to me (I don't understand the fast forward token part that well, so I did loose review on it btw)

**Comment by cadedaniel** [comment]
Can I review the fast forward part? It intersects with spec decode, and there's a few ways to implement it.

**Comment by cadedaniel** [comment]
On the topic of fast-forward tokens:

I agree we should support this optimization in vLLM, it's part of how SGLang's json mode has such good performance. However I feel we should decouple the implementation/design from this sampling controller pull request. There are design tradeoffs in how we implement fast-forward tokens, and we should consider the different ways before going with one.

Concretely, there's a different design that I believe is better for vLLM: use a specialized Worker which wraps the normal Worker, which contains the specialized logic which requires fast-forward tokens. For example, JSON mode could be implemented at as a JSONModeWorker, which wraps a normal worker within it.

The design benefit here is that from the wrapped Worker's perspective, jump tokens are specified using the same API as chunked prefill, speculative decoding scoring, and any other algorithms which require multiple query tokens per sequence (e.g. [classifier free guidance](https://github.com/vllm-project/vllm/issues/5825), or speculative edits ([fireworks blog post calls it "fast apply"](https://fireworks.ai/blog/cursor), [vllm issue](https://github.com/vllm-project/vllm/issues/5201))). Even if we just consider speculative decoding, we'd love to be able to combine speculative decoding with jump-tokens in JSON mode for extra speedup.

Two concrete downsides of the current design in this PR:
* It conflates scheduling status of a sequence with multi-query decode. We shouldn't need to make a sequence "PREFILL" to get multi-query decode, because that currently instructs the backend to use a different path for that kernel (e.g. fire off different attention kernels). For performance we should be able to have the backend fire off a single attention kernel which computes both multi-query and single-query attention scores. See this [design doc](https://docs.google.com/document/d/1I6H8bGYfwM8YappcP8HuoBhpS_pMlN0UIH2e_w9hv_M/edit#heading=h.smqr48km0plh) by @LiuXiaoxuanPKU for more information.
* It makes on-device code live outside of the Worker paradigm; e.g. a sampling controller that exists at the Engine level. To fully support different hardware backends and distributed configurations, we want to delegate the responsibility of all device code to below the Executor and Worker level. This can be trivially fixable in this PR by moving it to the Worker, but then we have the challenges of 1) the API only supports nvidia/amd backends, and 2) the interaction with Scheduler no longer works because the PR assumes worker 0 and engine share the same process. this is changing in [SPMD refactor](https://github.com/vllm-project/vllm/issues/6556) for other performance reasons.

**Comment by Harsha-Nori** [comment]
Hey all, I'm the lead maintainer for the Guidance project (https://github.com/guidance-ai/guidance), and we've had _many_ requests for guidance support on vLLM, which this PR would enable :). Guidance leverages "jump"/fast forward tokens in the majority of programs users write -- including many use cases beyond JSON mode -- so it'd be great for us to have infrastructure to leverage here in vLLM. Happy to help in any way I can to get this implemented!

**Comment by cadedaniel** [comment]
@mmoskal @Harsha-Nori I recommend following up with @zhuohan123 for best approach on jump tokens. the proposed approach has downsides for vLLM performance and maintanence.

**Comment by hmellor** [comment]
Hi @mmoskal do you plan to continue this work?

**Comment by mmoskal** [comment]
Not really given other efforts in the area. Thank for checking.

**Comment by simon-mo** [comment]
cc @njhill in which I believe you are looking into V1's sampler support.
